### ISSUE TITLE
Docker push should require a force flag to clobber

### ISSUE THREAD

**Issue by BMorearty** [docker/docker#2619]
If I push an image called bmorearty/xyz and then I make a brand new image (not based on the first one) and run `docker push bmorearty/xyz` again, the registry allows it.

To avoid unfortunate clobbering of images, this should not be allowed unless I pass in a "-force" option to the CLI.

**Comment by linquize** [comment]
Just like git, you have `git push --force`

**Comment by bojanstef** [comment]
#dibs this issue :+1:

**Comment by bojanstef** [comment]
I've been hacking away and I'm not sure if this issue still exists. 

To push an image one must run ```docker tag 7d9495d03763 username/xyz```
Then this will warn the user that an image with the same name already has that tag and to use the -f flag if they want to overwrite it. 

In other words, I can't seem to recreate your issue @BMorearty any words of wisdom?

**Comment by thaJeztah** [comment]
@bojanstef thanks for looking into this. I'm not sure if checking if an image exists locally (`docker tag`) is the right check, because even if an image doesn't exist locally, it can still exist in the registry that you'll push to. Also, there are plans to remove the `-f` option from `docker tag` because it's inconsistent with other commands; see https://github.com/docker/docker/issues/9798

If we want to implement this, we should either implement it in the registry (which will be a large breaking change), or we should check if an image with the same name/tag already exists in the registry before pushing, which may be a "race" and no 100% guarantee.

/cc @stevvooe @RichardScothern any thoughts? Should we try to implement this, or close as a "won't fix"?

**Comment by RichardScothern** [comment]
The registry API affords checking that an image/tag combination exists.  Given that, I'm loathe to make large breaking code and API changes.

**Comment by thaJeztah** [comment]
I'm not really comfortable with changing this either; this will be big breaking change, and will affect many users

**Comment by stevvooe** [comment]
@BMorearty @bojanstef While this is an interesting and useful feature, it will take a number of changes to the image format to realize correctly. For this particular issue, it seems like the actual problem is preventing accidental push over important images. This can be achieved with simple access control. Disallow push to your registry by everything except your CI system. This will avoid the problems with accidentally pushing and ensure the builds are reproducible.

Implementing this in the style of git is much more complex. In git, repositories must have an unambiguous, ordered set of changes that lead to a current HEAD for a given branch. A new HEAD must have the current head as a parent. Detecting if there is an inconsistency is as simple as verifying this property. In git parlance, we call this a "fast-forward". When you "git push --force", git skips this check and rewrites the head to whatever the user specifies.

This relationship is not at all true for docker images. This is due to limitations in the layer format. While git repositories may have thousands of commits leading to the current HEAD, a docker image is limited in the number of layers, usually by the performance of the underlying union filesystem.

Basically, adding this feature boils down to fixing the repository to a common parent. Given the current data model of docker images, this means that we can only add layers. Such images would be saddled by performance issues over time. Simple operations, such as replacing an operating system layer, involve a force push.

Now, this is not to say that such a feature is not a desire of docker image management. The first step in accomplishing this would be to represent changes to images in a way that doesn't involve creating a filesystem layer for each changeset. Once this is in place, one can then enforce the common parent constraint. When that is in place, it now makes sense to have a git-like force-push feature.

**Comment by keltieem** [comment]
@stevvooe  @thaJeztah Is this issue now closed?

**Comment by calavera** [comment]
There are several maintainers expressing their disconformity about this feature. I'm going to close this issue as "won't fix". Thanks a lot for your comments.

**Comment by olafure** [comment]
I do realise that "there are several maintainers expressing their disconformity about this feature", but please reconsider or consider an alternative:
One of the nature of versioning and a fundamental feature of a software registry is to know what you have, file what you have and reliably hand you back what you put in.  You put in version 1.2.3 and it will always give you that version back.

As for version, let's say 1.2.3, there is only one version 1.2.3.  There can never be an alternative version of your software with that version number.  If you ever think that would be a good idea, when you'll have to accept that you no longer have a versioned software.  You have something pretending to be versioning and can be really dangerous.

The only reason would ever have an alternative version of your software with that a colliding version number is due to a bug.  

I do understand that breaking changes are awful, but as an alternative, please consider making it an optional feature of the docker registry to prohibit (or requiring a flag) overwriting an existing version.

I'd also return a big red warning to the user who overwrites a version that this is a deprecated feature and that it'll require a special flag in future versions.

**Comment by thaJeztah** [comment]
@olafure the problem in this case is that `:tag` is a "free form" tag for images; docker, not the registry puts any meaning in it's use. So while one user may use tag to express a "semver" version of an image, another user can use if to, e.g. put a day in it (`myimage:monday`), and overwrite it each time.

The format does have _immutable_ tags (`image@sha256:deadbeef`) which are guaranteed to give the _exact_ same version of an image, and can be verified.

**Comment by olafure** [comment]
@thaJeztah I don't think that would be a problem, since this would be an optional configuration of the docker registry.  Those who wanted to be able to overwrite a tag could still do it.

To clear confusion, I wasn't suggesting any constraint on the tag name, "1.2.3" could as well be "monday-if-build-succeeded". 

The problem with immutable tag is:
- It's not the tag you specified
- It's not human readable 
- It doesn't tell you which tag you're pulling, just the checksum.

The major problem I'm trying to find a solution for is, that docker registry can't be trusted for the information you give it.  You give it a file and a tag, and you can't rely on it giving you the same file back unless you are willing to drop the tag and the information it contained.

**Comment by thaJeztah** [comment]
I _do_ see your point, but not sure it would really solve that, because as a _consumer_  of those images, you still wouldn't know if a tag was force-pushed, and the `:latest` tag is (by convention) used as a "rolling" tag. Possibly [Docker content trust](https://docs.docker.com/engine/security/trust/content_trust/) would resolve those issues for your use case?

**Comment by thaJeztah** [comment]
In any case; changing this would require this to be implemented in the _registry_ first, so it's probably best to discuss those options in that issue tracker; https://github.com/docker/distribution/issues
